### ISSUE TITLE
.NET Dynamic Providers

### ISSUE THREAD

**Issue by mikhailshilkov** [pulumi/pulumi#3638]

**Comment by Badabum** [comment]
Any plans on implementing this?

**Comment by mikhailshilkov** [comment]
@Badabum Dynamic providers aren't easily translatable to .NET the way they are implemented in node and python.

I think the in-progress implementation of #2430 can bring us a new way to implement dynamic providers, but I'd need to double-check with @pgavlin and @lukehoban  if that's correct.

**Comment by Badabum** [comment]
@mikhailshilkov thanks for the reply

**Comment by gitfool** [comment]
This keeps coming up as a crucial missing piece for .NET support. Holding my breath for a breakthrough. 🙏

**Comment by MisinformedDNA** [comment]
@mikhailshilkov What makes it not easily translatable? I found this [NodeJS](https://github.com/pulumi/pulumi/blob/master/sdk/nodejs/dynamic/index.ts) implementation, but it seems pretty straightforward. I'm sure I'm missing something.

**Comment by mikhailshilkov** [comment]
@MisinformedDNA Node.js dynamic providers are serialized and saved to the state files, which can't be easily copied to .NET.

Having said that, we have a working prototype of the .NET dynamic provider, so there are no blockers per se, just work to be done to bring that to the product.

**Comment by kralikba** [comment]
> 
> 
> @MisinformedDNA Node.js dynamic providers are serialized and saved to the state files, which can't be easily copied to .NET.
> 
> Having said that, we have a working prototype of the .NET dynamic provider, so there are no blockers per se, just work to be done to bring that to the product.

Is this prototype accessible? It would be really beneficial for us; we have a bunch of neat and not-so-neat hacks that could be thrown out and we could then have complete, idiomatic Pulumi config for our whole stack.

**Comment by mikhailshilkov** [comment]
I don't think it's currently accessible cc @justinvp

**Comment by kralikba** [comment]
would it be possible to get a sneak peek and maybe do some beta testing?

**Comment by gitfool** [comment]
Also keen to get my hands on this...

**Comment by tmarkovski** [comment]
Super excited about the .NET Dynamic Provider support! Very much needed for having clean infrastructure management!

**Comment by Jawvig** [comment]
Just adding my vote of support for this. Also, I'm very happy to try out a prototype and give feedback on it, if that helps.

**Comment by jmwaldrip** [comment]
Adding my vote of support for this. It is a crucial piece of .Net Support that I would love to get my hands on.

**Comment by TechWatching** [comment]
Hi. I was wondering what was the ETA on this.
I am using .NET to provision a SQL database with azure ad users and applications that have permissions on it properly configured and from what I understood I need to use a dynamic provider for that. @mikhailshilkov  does that mean I can't do it because dynamic providers are not yet available in .NET?

**Comment by mikhailshilkov** [comment]
@TechWatching Be sure to check if Azure AD or Azure Native ([SQL module](https://www.pulumi.com/registry/packages/azure-native/api-docs/sql/)) support what you need. But otherwise, yes, dynamic providers are the suggested workaround for missing resource types for TypeScript and Python.

.NET dynamic providers are still on the prototyping phase, I don't have a specific ETA to share.

**Comment by TechWatching** [comment]
@mikhailshilkov  Well the problem is that giving to permissions an azure ad user or an azure ad application on a sql database is done by executing an sql command so it's not handle directly by Azure APIs so not by Azure Native SQL Module (cf [ Microsoft documentation](https://docs.microsoft.com/fr-fr/azure/app-service/tutorial-connect-msi-sql-database?tabs=windowsclient%2Cdotnet#grant-permissions-to-managed-identity)). 

If it's not possible to do it with .NET dynamic providers, what should I use? Should I create a dynamic provider to do what I need using typescript, expose it as a Pulumi package that I can consume from my Pulumi program in .NET?

**Comment by Jawvig** [comment]
We've got exactly the same issue as @TechWatching. Would really rather not switch language, but will do if necessary and possible to use from .NET.

**Comment by mikhailshilkov** [comment]
One poor man's solution is to run your custom code in an `Apply` on an output of the target resource. This code will execute on every run (and every preview, but you can check that) after the resource is created, so you have to make it idempotent.

Another option is to implement a full-blown provider in Go ([example](https://github.com/mikhailshilkov/pulumi-provider-boilerplate)), but that's quite a time investment.

We'll share an update regarding .NET dynamic providers as soon as we have a runnable preview.

**Comment by Badabum** [comment]
@TechWatching you always can just add regular c# code(if what you need is possible in c#) to your pulumi program. you obviously lose diffs and all bells and whistles of pulumi but at least you don't need to wait for dynamic providers

**Comment by Badabum** [comment]
@UnoSD kudos for mentioning the lib! didn't know it exists

**Comment by ESipalis** [comment]
I'm looking for suggestions for generating some data for the initial creation of the resources, but not re-generating that data on every `pulumi up`. In other words, generate some values on the 1st `pulumi up` and then store them somewhere so they can be reused.

As an example, I need to generate an RSA key pair for CloudFront signed URLs. I use the public key in CloudFront public key's definition and the private key in Lambda Function's environment variables. If I just generate the keys using `RSA.Create()` and then pass them to the resources' inputs, those resources will be updated / replaced on every `pulumi up`. I can use `IgnoreChanges` as a workaround, but then if I add a third resource who uses one of the keys, the original resources will use the old key and the new resource will use a newly generated key.
I found out that I can use `Pulumi.Tls` package for my specific use case, but looking for a nice solution for possible future problems.
One idea, would be to check pulumi config and if that has no values set, generate them and save them into config. However, it's not possible to set config values in Pulumi programatically. Perhaps can just run `pulumi config set <key> <valye>` from .NET?
Another approach, I was thinking of was saving the values into Pulumi state using Dynamic providers, but that's also not supported in .NET.

So, now I'm thinking just generating the values and saving them into a simple file? Would be great to have pulumi integration with this though, to have built-in support for dealing with stacks and such.

**Comment by UnoSD** [comment]
> I'm looking for suggestions for generating some data for the initial creation of the resources, but not re-generating that data on every `pulumi up`. In other words, generate some values on the 1st `pulumi up` and then store them somewhere so they can be reused.
> 
> As an example, I need to generate an RSA key pair for CloudFront signed URLs. I use the public key in CloudFront public key's definition and the private key in Lambda Function's environment variables. If I just generate the keys using `RSA.Create()` and then pass them to the resources' inputs, those resources will be updated / replaced on every `pulumi up`. I can use `IgnoreChanges` as a workaround, but then if I add a third resource who uses one of the keys, the original resources will use the old key and the new resource will use a newly generated key. I found out that I can use `Pulumi.Tls` package for my specific use case, but looking for a nice solution for possible future problems. One idea, would be to check pulumi config and if that has no values set, generate them and save them into config. However, it's not possible to set config values in Pulumi programatically. Perhaps can just run `pulumi config set <key> <valye>` from .NET? Another approach, I was thinking of was saving the values into Pulumi state using Dynamic providers, but that's also not supported in .NET.
> 
> So, now I'm thinking just generating the values and saving them into a simple file? Would be great to have pulumi integration with this though, to have built-in support for dealing with stacks and such.

you can use Pulumi.Tls for generating the data only once and instead of using config for the value, I would just use outputs and you can check the previous stack outputs using `StackReference(Deployment.Instance.StackName).Outputs`

**Comment by ESipalis** [comment]
@UnoSD I figured the `Pulumi.Tls` approach. Just wondering how would I go about it if instead of RSA I'd need a random GUID for example. How I could achieve what `Pulumi.Tls` is doing, but with my own generated values instead of RSA keys.

**Comment by UnoSD** [comment]
Apologies to the others for going a little bit off topic here, but you can
use Pulumi.Random to generate reproducible pseudorandom guid

**Comment by bpolivka** [comment]
Any updates on this?

**Comment by Frassle** [comment]
Not really, this has been blocked on some other work and some internal discussion about how exactly to support it. I think all those issues are decided on now, so we should be able to make some more progress on this.
The first thing is we'll release support for writing providers for dotnet, then we'll be able to release dynamic providers on top of that.

**Comment by dferretti** [comment]
@Frassle writing providers in dotnet sounds very interesting! 👀 is there a github issue to subscribe to for that?

**Comment by Frassle** [comment]
Surprisingly not, one of our other engineers just opened https://github.com/pulumi/pulumi/issues/10915 for the same ask for Typescript so I've just opened for https://github.com/pulumi/pulumi/issues/10964 dotnet.
